src/core/openjpeg_decoder.js

```diff
@@ -0,0 +1,102 @@
+/* Copyright 2024 Mozilla Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { BaseException } from "../shared/util.js";
+import OpenJPEG from "../../external/openjpeg/openjpeg.js";
+
+class JpxError extends BaseException {
+  constructor(msg) {
+    super(`JPX error: ${msg}`, "JpxError");
+  }
+}
+
+class OpenJPEGDecoder {
+  #module = null;
+
+  static #instance = null;
+
+  constructor(module) {
+    this.#module = module;
+  }
+
+  static decode(data) {
+    this.#instance ||= new OpenJPEGDecoder(OpenJPEG());
+    return this.#instance.#decode(data);
+  }
+
+  #decode(data) {
+    const mod = this.#module;
+    const data_ptr = mod._malloc(data.length);
+    const image_ptr_ptr = mod._malloc(4);
+    const image_size_ptr = mod._malloc(4);
+
+    mod.HEAPU8.set(data, data_ptr);
+    const ret = mod._jp2_decode(
+      data_ptr,
+      data.length,
+      image_ptr_ptr,
+      image_size_ptr
+    );
+    mod._free(data_ptr);
+    if (ret) {
+      mod._free(image_ptr_ptr);
+      mod._free(image_size_ptr);
+      return new Uint8Array(0);
+    }
+
+    const image_ptr = mod.HEAPU32[image_ptr_ptr / 4];
+    const image_size = mod.HEAPU32[image_size_ptr / 4];
+    const newArray = mod.HEAPU8.slice(image_ptr, image_ptr + image_size);
+
+    mod._free(image_ptr);
+    mod._free(image_ptr_ptr);
+    mod._free(image_size_ptr);
+
+    return newArray;
```

Comment:
What sort of TypedArray is this?
To maintain consistency with the other image decoders, and the comment just below, this should be wrapped in an `Uint8ClampedArray` (unless it's already that format).